File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/CopySample/src/com/hp/workpath/sample/copysample/fragments/ShiftPreference.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.copysample.fragments;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.util.AttributeSet;

import androidx.preference.DialogPreference;
import androidx.preference.PreferenceManager;
import androidx.preference.PreferenceViewHolder;

import com.hp.workpath.sample.copysample.R;

public class ShiftPreference extends DialogPreference {

    float xShift = 0.0f;
    float yShift = 0.0f;

    float mXShiftMin = 0.0f;
    float mYShiftMin = 0.0f;
    float mXShiftMax = 0.0f;
    float mYShiftMax = 0.0f;

    public ShiftPreference (Context context, AttributeSet attrs, int defStyleAttr) { super(context, attrs, defStyleAttr); }

    public ShiftPreference (Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ShiftPreference (Context context) { super(context); }

    @Override
    public void onBindViewHolder(PreferenceViewHolder view) {
        super.onBindViewHolder(view);
        setDialogLayoutResource(R.layout.layout_shifts);
        loadShifts();
    }

    public void applyShifts() throws Exception {
        if (!isInRange(mXShiftMin, mXShiftMax, xShift)) {
            throw new Exception(getContext().getString(R.string.range_shift, mXShiftMin, mXShiftMax));
        }
        if (!isInRange(mYShiftMin, mYShiftMax, yShift)) {
            throw new Exception(getContext().getString(R.string.range_shift, mYShiftMin, mYShiftMax));
        }
        saveShifts();
    }

    private void loadShifts() {
        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(getContext());
        xShift = sharedPref.getFloat(getKey() + "xShift", 0.0f);
        yShift = sharedPref.getFloat(getKey() + "yShift", 0.0f);
        setShiftSummary();
    }

    private void setShiftSummary() {
        Handler handler = new Handler();
        final Runnable r = () -> setSummary(getContext().getString(R.string.summary_shift, xShift, yShift));
        handler.post(r);
    }

    private void saveShifts() {
        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(getContext());
        SharedPreferences.Editor editor = sharedPref.edit();
        editor.putFloat(getKey() + "xShift", xShift);
        editor.putFloat(getKey() + "yShift", yShift);
        editor.apply();
        setShiftSummary();
    }

    public void setXShiftLimits(final float min, final float max) {
        mXShiftMin = min;
        mXShiftMax = max;
    }

    public void setYShiftLimits(final float min, final float max) {
        mYShiftMin = min;
        mYShiftMax = max;
    }

    public void setShifts(float xShift, float yShift) {
        this.xShift = xShift;
        this.yShift = yShift;
    }

    public float getxShift() {
        return xShift;
    }

    public float getyShift() {
        return yShift;
    }

    private boolean isInRange(final float min, final float max, final float value) {
        return max > min ? value >= min && value <= max : value >= max && value <= min;
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/CopySample/src/com/hp/workpath/sample/copysample/fragments/ShiftPreference.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.copysample.fragments

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import androidx.preference.DialogPreference
import androidx.preference.PreferenceManager
import androidx.preference.PreferenceViewHolder
import com.hp.workpath.sample.copysample.R

class ShiftPreference : DialogPreference {
    var xShift = 0.0f
    var yShift = 0.0f
    private var mXShiftMin = 0.0f
    private var mYShiftMin = 0.0f
    private var mXShiftMax = 0.0f
    private var mYShiftMax = 0.0f

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {}
    constructor(context: Context?) : super(context) {}

    override fun onBindViewHolder(view: PreferenceViewHolder) {
        super.onBindViewHolder(view)
        dialogLayoutResource = R.layout.layout_shifts
        loadShifts()
    }

    @Throws(Throwable::class)
    fun applyShifts() {
        if (xShift !in mXShiftMin..mXShiftMax) {
            throw Throwable(context.getString(R.string.range_shift, mXShiftMin, mXShiftMax))
        }
        if (yShift !in mYShiftMin..mYShiftMax) {
            throw Throwable(context.getString(R.string.range_shift, mYShiftMin, mYShiftMax))
        }
        saveShifts()
    }

    private fun loadShifts() {
        val sharedPref = PreferenceManager.getDefaultSharedPreferences(context)
        xShift = sharedPref.getFloat(key + "xShift", 0.0f)
        yShift = sharedPref.getFloat(key + "yShift", 0.0f)
        setShiftSummary()
    }

    private fun setShiftSummary() {
        val handler = Handler(Looper.getMainLooper())
        val r = Runnable { summary = context.getString(R.string.summary_shift, xShift, yShift) }
        handler.post(r)
    }

    private fun saveShifts() {
        val sharedPref = PreferenceManager.getDefaultSharedPreferences(context)
        val editor = sharedPref.edit()
        editor.putFloat(key + "xShift", xShift)
        editor.putFloat(key + "yShift", yShift)
        editor.apply()
        setShiftSummary()
    }

    fun setXShiftLimits(min: Float, max: Float) {
        mXShiftMin = min
        mXShiftMax = max
    }

    fun setYShiftLimits(min: Float, max: Float) {
        mYShiftMin = min
        mYShiftMax = max
    }

    fun setShifts(xShift: Float, yShift: Float) {
        this.xShift = xShift
        this.yShift = yShift
    }
}